File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/fragment/NavDrawerFragment.java
/*
 * Copyright (c) 2016. Eli Connelly
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.emogoth.android.phone.mimi.fragment;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.text.format.DateUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.emogoth.android.phone.mimi.R;
import com.emogoth.android.phone.mimi.activity.MimiActivity;
import com.emogoth.android.phone.mimi.activity.StartupActivity;
import com.emogoth.android.phone.mimi.db.BoardTableConnection;
import com.emogoth.android.phone.mimi.db.DatabaseUtils;
import com.emogoth.android.phone.mimi.db.HistoryTableConnection;
import com.emogoth.android.phone.mimi.db.model.History;
import com.emogoth.android.phone.mimi.event.BookmarkClickedEvent;
import com.emogoth.android.phone.mimi.event.CloseTabEvent;
import com.emogoth.android.phone.mimi.event.HomeButtonPressedEvent;
import com.emogoth.android.phone.mimi.event.OpenHistoryEvent;
import com.emogoth.android.phone.mimi.event.UpdateHistoryEvent;
import com.emogoth.android.phone.mimi.fourchan.FourChanCommentParser;
import com.emogoth.android.phone.mimi.prefs.MimiSettings;
import com.emogoth.android.phone.mimi.util.BusProvider;
import com.emogoth.android.phone.mimi.util.LayoutType;
import com.emogoth.android.phone.mimi.util.MimiPrefs;
import com.emogoth.android.phone.mimi.util.MimiUtil;
import com.emogoth.android.phone.mimi.util.RxUtil;
import com.emogoth.android.phone.mimi.util.ThreadRegistry;
import com.emogoth.android.phone.mimi.view.DrawerViewHolder;
import com.mimireader.chanlib.models.ChanBoard;
import com.squareup.otto.Subscribe;

import java.util.List;

import rx.Subscription;
import rx.functions.Action1;


public class NavDrawerFragment extends Fragment {
    private static final String LOG_TAG = NavDrawerFragment.class.getSimpleName();

    private SharedPreferences sharedPreferences;

    private TextView loginRow;
    private int themeId;
    private int themeColorId;
    private FrameLayout notificationContainer;

    private MimiActivity activity;
    private View noBookmarksContainer;
    private ViewGroup bookmarksItemContainer;
    private List<History> currentBookmarks;
    private int fontSizeId;
    private String layoutType;

    private Subscription boardInfoSubscription;
    private Subscription bookmarkSubscription;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        setRetainInstance(false);

        activity = (MimiActivity) getActivity();

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        themeId = Integer.valueOf(sharedPreferences.getString(getString(R.string.theme_pref), "0"));
        themeColorId = Integer.valueOf(sharedPreferences.getString(getString(R.string.theme_color_pref), "0"));
        fontSizeId = Integer.valueOf(sharedPreferences.getString(getString(R.string.font_style_pref), "0"));
        layoutType = sharedPreferences.getString(getString(R.string.start_activity_pref), StartupActivity.getDefaultStartupActivity());

        final View v = inflater.inflate(R.layout.fragment_nav_drawer, container, false);
        return v;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final View settingsRow = view.findViewById(R.id.settings_row);
        settingsRow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivityForResult(new Intent(getActivity(), MimiSettings.class), MimiActivity.SETTINGS_ID);
                toggleDrawer();
            }

        });

        notificationContainer = (FrameLayout) view.findViewById(R.id.notification_container);
        notificationContainer.addView(MimiUtil.getInstance().createActionBarNotification(getActivity().getLayoutInflater(), notificationContainer, ThreadRegistry.getInstance().getUnreadCount()));

        final View homeRow = view.findViewById(R.id.home_row);
        homeRow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                BusProvider.getInstance().post(new HomeButtonPressedEvent());
                toggleDrawer();
            }
        });

        final View closeTabsContainer = view.findViewById(R.id.close_tabs_container);
        if (MimiUtil.getLayoutType(getActivity()) != LayoutType.TABBED) {
            closeTabsContainer.setVisibility(View.GONE);
        } else {
            final View closeTabsRow = view.findViewById(R.id.close_tabs_row);
            closeTabsRow.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    closeAllTabs();
                }
            });
        }

        final View bookmarksRow = view.findViewById(R.id.bookmarks_row);
        bookmarksRow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toggleDrawer();
                BusProvider.getInstance().post(new OpenHistoryEvent(true));
            }
        });

        bookmarksItemContainer = (ViewGroup) view.findViewById(R.id.bookmark_items);
        noBookmarksContainer = view.findViewById(R.id.no_bookmarks);
        populateBookmarks();

        final View historyRow = view.findViewById(R.id.history_row);
        historyRow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toggleDrawer();
                BusProvider.getInstance().post(new OpenHistoryEvent(false));
            }
        });
    }

    private void toggleDrawer() {
        if (getActivity() instanceof MimiActivity) {
            ((MimiActivity) getActivity()).toggleNavDrawer();
        }
    }

    private void closeAllTabs() {
        final LayoutInflater inflater = LayoutInflater.from(getActivity());
        final android.app.AlertDialog.Builder dialogBuilder = new android.app.AlertDialog.Builder(getActivity());
        final View dialogView = inflater.inflate(R.layout.dialog_close_tabs_prompt, null, false);
        final CheckBox dontShow = (CheckBox) dialogView.findViewById(R.id.dialog_dont_show);
        final SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(getActivity());
        final boolean shouldShowDialog = pref.getBoolean(getString(R.string.close_all_tabs_prompt_pref), true);

        if (shouldShowDialog) {
            dialogBuilder.setTitle(R.string.close_all_tabs)
                    .setView(dialogView)
                    .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            final SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(getActivity());
                            pref.edit().putBoolean(getString(R.string.close_all_tabs_prompt_pref), !dontShow.isChecked()).apply();

                            BusProvider.getInstance().post(new CloseTabEvent(true));
                            toggleDrawer();
                        }
                    })
                    .setNegativeButton(R.string.no, null)
                    .show();
        } else {
            BusProvider.getInstance().post(new CloseTabEvent(true));
            toggleDrawer();
        }

    }

    public void populateBookmarks() {
        final int bookmarkCount = MimiPrefs.navDrawerBookmarkCount(getActivity());

//        HistoryTableConnection.fetchPost(12345)
//                .subscribe();
        RxUtil.safeUnsubscribe(bookmarkSubscription);
        bookmarkSubscription = HistoryTableConnection.fetchHistory(true, bookmarkCount)
                .subscribe(new Action1<List<History>>() {
                    @Override
                    public void call(List<History> bookmarks) {
                        currentBookmarks = bookmarks;

                        bookmarksItemContainer.removeAllViews();
                        bookmarksItemContainer.addView(noBookmarksContainer);

                        if (bookmarks != null && bookmarks.size() > 0) {
                            notificationContainer.setVisibility(View.VISIBLE);
                            noBookmarksContainer.setVisibility(View.GONE);
                            for (int i = 0; i < bookmarks.size(); i++) {
                                final History bookmark = bookmarks.get(i);
                                bookmark.orderId = i;

                                final View row = getActivity().getLayoutInflater().inflate(R.layout.bookmark_row_item, bookmarksItemContainer, false);
                                row.setOnClickListener(new View.OnClickListener() {
                                    @Override
                                    public void onClick(View v) {
                                        sendBookmarkClickedEvent(bookmark);
                                        toggleDrawer();
                                    }
                                });
                                bookmarksItemContainer.addView(createBookmarkRow(bookmark, row));
                            }
                        } else {
                            notificationContainer.setVisibility(View.INVISIBLE);
                            noBookmarksContainer.setVisibility(View.VISIBLE);
                        }
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        Log.e(LOG_TAG, "Error loading bookmarks", throwable);
                    }
                });
    }

    private void sendBookmarkClickedEvent(final History bookmark) {
        RxUtil.safeUnsubscribe(boardInfoSubscription);
        boardInfoSubscription = BoardTableConnection.fetchBoard(bookmark.boardName)
                .compose(DatabaseUtils.<ChanBoard>applySchedulers())
                .subscribe(new Action1<ChanBoard>() {
                    @Override
                    public void call(ChanBoard chanBoard) {
                        if (chanBoard != null) {
                            BusProvider.getInstance().post(new BookmarkClickedEvent(bookmark.threadId, chanBoard.getName(), chanBoard.getTitle(), bookmark.orderId));
                        } else if (getActivity() != null) {
                            Toast.makeText(getActivity(), R.string.error_occurred, Toast.LENGTH_SHORT).show();
                        }
                    }
                });
    }

    private View createBookmarkRow(final History bookmark, final View row) {

        final CharSequence time = DateUtils.getRelativeTimeSpanString(
                bookmark.lastAccess,
                System.currentTimeMillis(),
                DateUtils.MINUTE_IN_MILLIS,
                DateUtils.FORMAT_ABBREV_RELATIVE);
        final DrawerViewHolder viewHolder = new DrawerViewHolder(row);
        final FourChanCommentParser.Builder parserBuilder = new FourChanCommentParser.Builder();
        parserBuilder.setContext(getActivity())
                .setComment(bookmark.text)
                .setBoardName(bookmark.boardName)
                .setThreadId(bookmark.threadId)
                .setQuoteColor(MimiUtil.getInstance().getQuoteColor())
                .setReplyColor(MimiUtil.getInstance().getReplyColor())
                .setHighlightColor(MimiUtil.getInstance().getHighlightColor())
                .setLinkColor(MimiUtil.getInstance().getLinkColor());

        viewHolder.text.setText(parserBuilder.build().parse());

        if (bookmark.watched) {
            final int count = ThreadRegistry.getInstance().getUnreadCount(bookmark.threadId);

            if (count > 0) {
                viewHolder.unreadcount.setText(String.valueOf(count));
                viewHolder.unreadcount.setVisibility(View.VISIBLE);
            } else {
                viewHolder.unreadcount.setVisibility(View.GONE);
            }
        } else {
            viewHolder.unreadcount.setVisibility(View.GONE);
        }

        viewHolder.image.setVisibility(View.INVISIBLE);
        if (!TextUtils.isEmpty(bookmark.tim)) {
            final String url = MimiUtil.httpOrHttps(activity) + activity.getString(R.string.thumb_link) + activity.getString(R.string.thumb_path, bookmark.boardName, bookmark.tim);

            Glide.with(activity)
                    .load(url)
                    .crossFade()
                    .error(R.drawable.placeholder_image)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(viewHolder.image);

            viewHolder.image.setVisibility(View.VISIBLE);

//            viewHolder.image.setDefaultImageResId(R.drawable.ic_content_picture);
//            viewHolder.image.setErrorImageResId(R.drawable.ic_content_picture);
//            viewHolder.image.setImageUrl(url, imageLoader);

        } else {
            viewHolder.image.setVisibility(View.INVISIBLE);
            Glide.clear(viewHolder.image);
        }

        viewHolder.boardName.setText("/" + bookmark.boardName + "/");
        viewHolder.threadId.setText(String.valueOf(bookmark.threadId));
        viewHolder.lastviewed.setText(time);

        return row;
    }

//    @Subscribe
//    public void openThreadInActivity(final HistoryDbModel historyItem) {
//        final Bundle args = new Bundle();
//        final Class clazz;
//        args.putBoolean(Extras.EXTRAS_USE_BOOKMARKS, true);
//        args.putInt(Extras.EXTRAS_VIEWING_HISTORY, MimiActivity.VIEWING_BOOKMARKS);
//
//        args.putString(Extras.EXTRAS_BOARD_NAME, historyItem.getBoardName());
//        args.putInt(Extras.EXTRAS_THREAD_ID, historyItem.getThreadId());
//        args.putInt(Extras.EXTRAS_POSITION, historyItem.getOrderId());
//
//        final List<HistoryDbModel> bookmarks = MimiUtil.getInstance().getBookmarks();
//        final ArrayList<ThreadInfo> threadList = new ArrayList<>(bookmarks.size());
//
//        for (final HistoryDbModel post : bookmarks) {
//            final ThreadInfo threadInfo = new ThreadInfo(post.getThreadId(), post.getBoardName(), null);
//            threadList.add(threadInfo);
//        }
//
//        args.putParcelableArrayList(Extras.EXTRAS_THREAD_LIST, threadList);
//
//        if (getResources().getBoolean(R.bool.two_pane)) {
//            clazz = PostItemListActivity.class;
//        } else {
//            clazz = PostItemDetailActivity.class;
//        }
//
//        toggleDrawer();
//
//        final Intent intent = new Intent(getActivity(), clazz);
//        intent.putExtras(args);
//        startActivity(intent);
//
//    }

    @Override
    public void onPause() {
        super.onPause();

        BusProvider.getInstance().unregister(this);
        RxUtil.safeUnsubscribe(boardInfoSubscription);
    }

    @Override
    public void onResume() {
        super.onResume();
        BusProvider.getInstance().register(this);

        final int themePref = Integer.valueOf(sharedPreferences.getString(getString(R.string.theme_pref), "0"));
        final int themeColorPref = Integer.valueOf(sharedPreferences.getString(getString(R.string.theme_color_pref), "0"));

        if (themeId != themePref || themeColorPref != themeColorId) {
            themeId = themePref;
            themeColorId = themeColorPref;

            MimiUtil.getInstance().setCurrentTheme(themePref, themeColorPref);
            Activity activity = getActivity();
            Intent intent = new Intent(activity, activity.getClass());
            activity.finish();
            startActivity(intent);

        }

        final int fontSizePref = Integer.valueOf(sharedPreferences.getString(getString(R.string.font_style_pref), "0"));
        if (fontSizeId != fontSizePref) {
            Activity activity = getActivity();
            Intent intent = new Intent(activity, activity.getClass());
            activity.finish();
            startActivity(intent);
        }

        final String layoutTypePref = sharedPreferences.getString(getString(R.string.start_activity_pref), StartupActivity.getDefaultStartupActivity());
        if (!TextUtils.equals(layoutType, layoutTypePref)) {
            Activity activity = getActivity();
            Intent intent = new Intent(activity, StartupActivity.class);
            activity.finish();
            startActivity(intent);
        }

        if (loginRow != null && MimiUtil.getInstance().isLoggedIn()) {
            loginRow.setText(R.string.chanpass_logout);
        }

    }

    @Subscribe
    public void onAutoRefresh(final UpdateHistoryEvent event) {
        populateBookmarks();

        notificationContainer.removeAllViews();
        notificationContainer.addView(MimiUtil.getInstance().createActionBarNotification(getActivity().getLayoutInflater(), notificationContainer, ThreadRegistry.getInstance().getUnreadCount()));
    }

}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/fragment/NavDrawerFragment.kt
package com.emogoth.android.phone.mimi.fragment

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.text.format.DateUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.FrameLayout
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.preference.PreferenceManager
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.emogoth.android.phone.mimi.R
import com.emogoth.android.phone.mimi.activity.MimiActivity
import com.emogoth.android.phone.mimi.activity.StartupActivity
import com.emogoth.android.phone.mimi.db.BoardTableConnection.fetchBoard
import com.emogoth.android.phone.mimi.db.DatabaseUtils.applySchedulers
import com.emogoth.android.phone.mimi.db.DatabaseUtils.applySingleSchedulers
import com.emogoth.android.phone.mimi.db.HistoryTableConnection.fetchHistory
import com.emogoth.android.phone.mimi.db.HistoryTableConnection.fetchPost
import com.emogoth.android.phone.mimi.db.HistoryTableConnection.observeHistory
import com.emogoth.android.phone.mimi.db.models.History
import com.emogoth.android.phone.mimi.fourchan.FourChanCommentParser
import com.emogoth.android.phone.mimi.interfaces.HistoryClickedListener
import com.emogoth.android.phone.mimi.interfaces.HomeButtonListener
import com.emogoth.android.phone.mimi.interfaces.TabEventListener
import com.emogoth.android.phone.mimi.prefs.SettingsActivity
import com.emogoth.android.phone.mimi.util.GlideApp
import com.emogoth.android.phone.mimi.util.LayoutType
import com.emogoth.android.phone.mimi.util.MimiPrefs.Companion.navDrawerBookmarkCount
import com.emogoth.android.phone.mimi.util.MimiUtil
import com.emogoth.android.phone.mimi.util.RxUtil
import com.emogoth.android.phone.mimi.view.DrawerViewHolder
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mimireader.chanlib.models.ChanBoard
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

class NavDrawerFragment : Fragment() {
    //    private var sharedPreferences: SharedPreferences? = null
    private val loginRow: TextView? = null
    private var themeId = 0
    private var themeColorId = 0
    private var notificationContainer: FrameLayout? = null
    private var noBookmarksContainer: View? = null
    private var bookmarksItemContainer: ViewGroup? = null
    private var currentBookmarks: List<History>? = null
    private var fontSizeId = 0
    private var layoutType: String? = null
    private var boardInfoSubscription: Disposable? = null
    private var bookmarkSubscription: Disposable? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        retainInstance = false
        if (activity == null) {
            return null
        }

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity)
        themeId = sharedPreferences.getString(getString(R.string.theme_pref), "0")?.toInt() ?: 0
        themeColorId = sharedPreferences.getString(getString(R.string.theme_color_pref), "0")?.toInt()
                ?: 0
        fontSizeId = sharedPreferences.getString(getString(R.string.font_style_pref), "0")?.toInt()
                ?: 0
        layoutType = sharedPreferences.getString(getString(R.string.start_activity_pref), StartupActivity.getDefaultStartupActivity())
        return inflater.inflate(R.layout.fragment_nav_drawer, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val settingsRow = view.findViewById<View>(R.id.settings_row)
        settingsRow.setOnClickListener { v: View? ->
            startActivityForResult(Intent(activity, SettingsActivity::class.java), MimiActivity.SETTINGS_ID)
            toggleDrawer()
        }
        notificationContainer = view.findViewById(R.id.notification_container)
        totalUnread(object : UnreadCountUpdate {
            override fun OnUnreadCountUpdate(count: Int) {
                if (activity != null) {
                    val act = activity as MimiActivity
                    val nc = notificationContainer
                    nc?.addView(MimiUtil.getInstance().createActionBarNotification(act.layoutInflater, nc, count))
                }
            }
        })
        val homeRow = view.findViewById<View>(R.id.home_row)
        homeRow.setOnClickListener { v: View? ->
            if (activity is HomeButtonListener) {
                (activity as HomeButtonListener).onHomeButtonClicked()
            }

            toggleDrawer()
        }
        val closeTabsContainer = view.findViewById<View>(R.id.close_tabs_container)
        if (MimiUtil.getLayoutType(activity) != LayoutType.TABBED) {
            closeTabsContainer.visibility = View.GONE
        } else {
            val closeTabsRow = view.findViewById<View>(R.id.close_tabs_row)
            closeTabsRow.setOnClickListener { v: View? -> closeAllTabs() }
        }
        val bookmarksRow = view.findViewById<View>(R.id.bookmarks_row)
        bookmarksRow.setOnClickListener { v: View? ->
            toggleDrawer()
            if (activity is HistoryClickedListener) {
                val listener = activity as HistoryClickedListener
                listener.openHistoryPage(true)
            }
        }
        bookmarksItemContainer = view.findViewById(R.id.bookmark_items)
        noBookmarksContainer = view.findViewById(R.id.no_bookmarks)
        val historyRow = view.findViewById<View>(R.id.history_row)
        historyRow.setOnClickListener { v: View? ->
            toggleDrawer()
            if (activity is HistoryClickedListener) {
                val listener = activity as HistoryClickedListener
                listener.openHistoryPage(false)
            }
        }

    }

    private fun toggleDrawer() {
        if (activity is MimiActivity) {
            (activity as MimiActivity).toggleNavDrawer()
        }
    }

    private fun closeAllTabs() {
        if (activity == null) {
            return
        }

        val inflater = LayoutInflater.from(activity)
        val dialogBuilder = MaterialAlertDialogBuilder(activity as Context)
        val dialogView = inflater.inflate(R.layout.dialog_close_tabs_prompt, null, false)
        val dontShow = dialogView.findViewById<View>(R.id.dialog_dont_show) as CheckBox
        val pref = PreferenceManager.getDefaultSharedPreferences(activity)
        val shouldShowDialog = pref.getBoolean(getString(R.string.close_all_tabs_prompt_pref), true)
        if (shouldShowDialog) {
            dialogBuilder.setTitle(R.string.close_all_tabs)
                    .setView(dialogView)
                    .setPositiveButton(R.string.yes) { dialog, which ->
                        pref.edit().putBoolean(getString(R.string.close_all_tabs_prompt_pref), !dontShow.isChecked).apply()
                        if (activity is TabEventListener) {
                            val act = activity as TabEventListener
                            act.onTabClosed(-1, "", "", true)
                        }

                        toggleDrawer()
                    }
                    .setNegativeButton(R.string.no, null)
                    .show()
        } else if (activity is TabEventListener) {
            val act = activity as TabEventListener
            act.onTabClosed(-1, "", "", true)

            toggleDrawer()
        }
    }

    private fun populateBookmarks() {
        if (activity == null || !isAdded) {
            return
        }
        val bookmarkCount = navDrawerBookmarkCount(activity as Context)
        RxUtil.safeUnsubscribe(bookmarkSubscription)
        bookmarkSubscription = observeHistory(true, bookmarkCount)
                .compose(applySchedulers())
                .subscribe({ bookmarks: List<History> ->
                    if (activity == null || !isAdded) {
                        return@subscribe
                    }
                    currentBookmarks = bookmarks
                    bookmarksItemContainer?.removeAllViews()
                    bookmarksItemContainer?.addView(noBookmarksContainer)
                    if (bookmarks.isNotEmpty()) {
                        notificationContainer?.visibility = View.VISIBLE
                        noBookmarksContainer?.visibility = View.GONE
                        for (i in bookmarks.indices) {
                            val bookmark = bookmarks[i]
                            bookmark.orderId = i
                            val row = (activity as FragmentActivity).layoutInflater.inflate(R.layout.bookmark_row_item, bookmarksItemContainer, false)
                            row.setOnClickListener { v: View? ->
                                sendBookmarkClickedEvent(bookmark)
                                toggleDrawer()
                            }
                            bookmarksItemContainer?.addView(createBookmarkRow(bookmark, row))
                        }
                    } else {
                        notificationContainer?.visibility = View.INVISIBLE
                        noBookmarksContainer?.visibility = View.VISIBLE
                    }

                    var unread = 0
                    for (history in bookmarks) {
                        unread += history.unreadCount
                    }

                    val act = activity as MimiActivity
                    val nc = notificationContainer
                    nc?.addView(MimiUtil.getInstance().createActionBarNotification(act.layoutInflater, nc, unread))


                }) { throwable: Throwable? -> Log.e(LOG_TAG, "Error loading bookmarks", throwable) }
    }

    private fun sendBookmarkClickedEvent(bookmark: History) {
        RxUtil.safeUnsubscribe(boardInfoSubscription)
        boardInfoSubscription = fetchBoard(bookmark.boardName)
                .compose(applySingleSchedulers())
                .subscribe { chanBoard: ChanBoard ->
                    if (!TextUtils.isEmpty(chanBoard.name)) {
                        if (activity is HistoryClickedListener) {
                            val act = activity as HistoryClickedListener
                            act.onHistoryItemClicked(chanBoard.name, bookmark.threadId, chanBoard.title, bookmark.orderId, bookmark.watched)
                        }
                    } else if (activity != null) {
                        Toast.makeText(activity, R.string.error_occurred, Toast.LENGTH_SHORT).show()
                    }
                }
    }

    private fun createBookmarkRow(bookmark: History, row: View): View {
        val time = DateUtils.getRelativeTimeSpanString(
                bookmark.lastAccess,
                System.currentTimeMillis(),
                DateUtils.MINUTE_IN_MILLIS,
                DateUtils.FORMAT_ABBREV_RELATIVE)
        val viewHolder = DrawerViewHolder(row)
        val parserBuilder = FourChanCommentParser.Builder()
        parserBuilder.setContext(activity)
                .setComment(bookmark.text)
                .setBoardName(bookmark.boardName)
                .setThreadId(bookmark.threadId)
                .setQuoteColor(MimiUtil.getInstance().quoteColor)
                .setReplyColor(MimiUtil.getInstance().replyColor)
                .setHighlightColor(MimiUtil.getInstance().highlightColor)
                .setLinkColor(MimiUtil.getInstance().linkColor)
        viewHolder.text.text = parserBuilder.build().parse()
        if (bookmark.watched) {
            if (bookmark.unreadCount > 0) {
                viewHolder.unreadcount.text = bookmark.unreadCount.toString()
                viewHolder.unreadcount.visibility = View.VISIBLE
            } else {
                viewHolder.unreadcount.visibility = View.GONE
            }
        } else {
            viewHolder.unreadcount.visibility = View.GONE
        }
        viewHolder.image.visibility = View.INVISIBLE
        if (!TextUtils.isEmpty(bookmark.tim)) {
            val thumbBaseUrl = activity?.getString(R.string.thumb_link)
            val thumbPath = activity?.getString(R.string.thumb_path, bookmark.boardName, bookmark.tim)
            val url = MimiUtil.https() + thumbBaseUrl + thumbPath
            GlideApp.with(activity as Activity)
                    .load(url)
                    .error(R.drawable.placeholder_image)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(viewHolder.image)
            viewHolder.image.visibility = View.VISIBLE
        } else {
            viewHolder.image.visibility = View.INVISIBLE
            GlideApp.with(activity as Activity).clear(viewHolder.image)
        }
        viewHolder.boardName.text = "/" + bookmark.boardName + "/"
        viewHolder.threadId.text = bookmark.threadId.toString()
        viewHolder.lastviewed.text = time
        return row
    }

    override fun onPause() {
        super.onPause()
        RxUtil.safeUnsubscribe(bookmarkSubscription)
        RxUtil.safeUnsubscribe(boardInfoSubscription)
    }

    override fun onResume() {
        super.onResume()

        populateBookmarks()

        if (activity == null) {
            return
        }
        val act = activity as Activity

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity)
        val themePref = sharedPreferences.getString(getString(R.string.theme_pref), "0")?.toInt()
                ?: 0
        val themeColorPref = sharedPreferences.getString(getString(R.string.theme_color_pref), "0")?.toInt()
                ?: 0
        if (themeId != themePref || themeColorPref != themeColorId) {
            themeId = themePref
            themeColorId = themeColorPref
            MimiUtil.getInstance().setCurrentTheme(themePref, themeColorPref)
            val intent = Intent(activity, act.javaClass)
            act.finish()
            startActivity(intent)

        }
        val fontSizePref = sharedPreferences.getString(getString(R.string.font_style_pref), "0")?.toInt()
                ?: 0
        if (fontSizeId != fontSizePref) {
            val intent = Intent(act, act.javaClass)
            act.finish()
            startActivity(intent)
        }
        val layoutTypePref = sharedPreferences.getString(getString(R.string.start_activity_pref), StartupActivity.getDefaultStartupActivity())
        if (!TextUtils.equals(layoutType, layoutTypePref)) {
            val intent = Intent(act, StartupActivity::class.java)
            act.finish()
            startActivity(intent)
        }
        if (loginRow != null && MimiUtil.getInstance().isLoggedIn) {
            loginRow.setText(R.string.chanpass_logout)
        }
    }

    private fun postUnread(boardName: String, threadId: Long, callback: UnreadCountUpdate) {
        fetchPost(boardName, threadId)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<History> {
                    override fun onSubscribe(d: Disposable) {
                        // no op
                    }

                    override fun onSuccess(history: History) {
                        callback.OnUnreadCountUpdate(history.unreadCount)
                    }

                    override fun onError(e: Throwable) {
                        Log.e(LOG_TAG, "Error setting unread count badge", e)
                    }
                })
    }

    private fun totalUnread(callback: UnreadCountUpdate) {
        fetchHistory(true)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<List<History>> {
                    override fun onSubscribe(d: Disposable) {
                        // no op
                    }

                    override fun onSuccess(histories: List<History>) {
                        var unread = 0
                        for (history in histories) {
                            unread += history.unreadCount
                        }
                        callback.OnUnreadCountUpdate(unread)
                    }

                    override fun onError(e: Throwable) {
                        Log.e(LOG_TAG, "Error setting unread count badge", e)
                    }
                })
    }

    override fun onDestroy() {
        super.onDestroy()
        RxUtil.safeUnsubscribe(bookmarkSubscription)
        RxUtil.safeUnsubscribe(boardInfoSubscription)
    }

    private interface UnreadCountUpdate {
        fun OnUnreadCountUpdate(count: Int)
    }

    companion object {
        private val LOG_TAG = NavDrawerFragment::class.java.simpleName
    }
}